sdk/servicebus/service-bus/src/core/shared.ts

```diff
@@ -103,3 +106,127 @@ export function createReceiverOptions(
 
   return rcvrOptions;
 }
+
+export const UnsettledMessagesLimitExceededError =
+  "Failed to fetch new messages as the limit for unsettled messages is reached. Please settle received messages using settlement methods(such as `completeMessage()`) on the receiver to receive the next message.";
+
+/**
+ * Returns the number of empty/filled slots in the Circular buffer of incoming deliveries
+ * based on the capacity and size of the buffer.
+ *
+ * @internal
+ */
+export function incomingBufferProperties(
+  receiver: Pick<Receiver, "session"> | undefined
+): {
+  numberOfEmptySlots: number; // 2048(total) - filled
+  numberOfFilledSlots: number; // number of unsettled messages
+} {
+  const incomingDeliveries = receiver?.session?.incoming?.deliveries;
+  return {
+    numberOfEmptySlots: incomingDeliveries
+      ? incomingDeliveries.capacity - incomingDeliveries.size - 1
```

Comment:
Is this meant to be this?
```suggestion
      ? incomingDeliveries.capacity - (incomingDeliveries.size - 1)
```
Otherwise if capacity and size are ever equal, you'll end up with `-1` as the `numberOfEmptySlots`. 

I'm not sure why there's a `-1`. If capacity is 2048, and size (I presume this is the number of messages) is 2048, wouldn't we expect 0 empty slots?

Reply:
Also, thanks for the comment. Capacity and size are never equal because I have -1.

2048 is when we run into problems.. with the above code, I'm only ever giving a max of 2047 empty slots to avoid the edge case(and the problems with the service).

Initial 
- size = 0
- capacity = 2048

Final
- size =2047
- capacity = 2048

This is never negative. (If it helps, I can add code to make sure this is never negative.)

That said, I may not need -1, I'll re-test this without -1 to see if I need it and add comments as needed.

Reply:
> Capacity and size are never equal because I have -1.
But without the parenthesis, the logic would be:
`incomingDeliveries.capacity - incomingDeliveries.size - 1` = `2048 - 2048 - 1` = `0 - 1` = `-1`.

What am I missing?

Reply:
> Edit: Is there some other logic that ensures incomingDeliveries.size never reaches 2048?

Size is basically the number of messages in the buffer, which is again based on the initial state. 
Whenever we add credits to the link, we make sure it doesn't exceed the emptySlots.
And the max emptySlots can only be 2048 - 1(initial).
So, size can only reach to a max of 2048 - 1.

Reply:
I just repro-ed, the exact problem is that, if I allow the size(number of messages) to be 2048, consequent drain gets stuck.
So, I need -1.

Reply:
9bf59cbe1f4f80fc2e8a3fe31e91b6fb9d305344